gax/src/main/java/com/google/api/gax/rpc/RetryingCallable.java

```diff
@@ -29,36 +29,52 @@
  */
 package com.google.api.gax.rpc;
 
+import com.google.api.core.ApiClock;
+import com.google.api.gax.retrying.ExponentialRetryAlgorithm;
+import com.google.api.gax.retrying.RetryAlgorithm;
+import com.google.api.gax.retrying.RetrySettings;
 import com.google.api.gax.retrying.RetryingExecutor;
 import com.google.api.gax.retrying.RetryingFuture;
+import com.google.api.gax.retrying.ScheduledRetryingExecutor;
 import com.google.common.base.Preconditions;
+import java.util.concurrent.ScheduledExecutorService;
 
 /**
  * A UnaryCallable that will keep issuing calls to an inner callable until it succeeds or times out.
  *
  * <p>Package-private for internal use.
  */
 class RetryingCallable<RequestT, ResponseT> extends UnaryCallable<RequestT, ResponseT> {
-  private final ApiCallContext callContextPrototype;
   private final UnaryCallable<RequestT, ResponseT> callable;
-  private final RetryingExecutor<ResponseT> executor;
+  private final ScheduledExecutorService executor;
+  private final RetrySettings retrySettings;
+  private final ApiClock clock;
 
-  RetryingCallable(
-      ApiCallContext callContextPrototype,
-      UnaryCallable<RequestT, ResponseT> callable,
-      RetryingExecutor<ResponseT> executor) {
-    this.callContextPrototype = Preconditions.checkNotNull(callContextPrototype);
-    this.callable = Preconditions.checkNotNull(callable);
-    this.executor = Preconditions.checkNotNull(executor);
+  RetryingCallable(UnaryCallable<RequestT, ResponseT> innerCallable,
+      RetrySettings retrySettings, ApiClock clock,
+      ScheduledExecutorService executor) {
+
+    this.callable = innerCallable;
+    this.retrySettings = retrySettings;
+    this.clock = clock;
+    this.executor = executor;
   }
 
   @Override
-  public RetryingFuture<ResponseT> futureCall(RequestT request, ApiCallContext inputContext) {
-    ApiCallContext context = callContextPrototype.nullToSelf(inputContext);
+  public RetryingFuture<ResponseT> futureCall(RequestT request, ApiCallContext context) {
+    RetryAlgorithm<ResponseT> retryAlgorithm =
```

Comment:
Maybe the right "cut point" is to pass in a `RetryAlgorithm` instead of `RetrySettings` and `ApiClock`

Reply:
I don't think it matters for this feature, but I think this will be needed for the feature request in #561. To allow callers to configure an overall deadline for the operation in the call context, the retry algorithm will need access to overall timeout set in the call context. So the retry algorithm will need to be instantiated in futureCall.
I haven't fully thought this through, so if you prefer I can raise the cut point util I start working on that feature. But if you don't have strong feelings, I'd prefer to keep it as is